File: dev-settings/coding-agents/sync-cmd/internal/checksum/checksum.go
package checksum

import (
	"crypto/sha256"
	"fmt"
	"io"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
)

func fileDigest(path string) string {
	f, err := os.Open(path)
	if err != nil {
		return "NOT_EXISTS"
	}
	defer f.Close()

	h := sha256.New()
	if _, err := io.Copy(h, f); err != nil {
		return "NOT_EXISTS"
	}
	return fmt.Sprintf("%x", h.Sum(nil))
}

func directoryDigest(dirPath string) string {
	info, err := os.Stat(dirPath)
	if err != nil || !info.IsDir() {
		return "NOT_EXISTS"
	}

	h := sha256.New()

	type entry struct {
		relPath string
		absPath string
	}
	var entries []entry

	filepath.WalkDir(dirPath, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() {
			return nil
		}
		rel, _ := filepath.Rel(dirPath, path)
		entries = append(entries, entry{relPath: rel, absPath: path})
		return nil
	})

	sort.Slice(entries, func(i, j int) bool {
		return entries[i].relPath < entries[j].relPath
	})

	for _, e := range entries {
		h.Write([]byte(e.relPath))
		h.Write([]byte(fileDigest(e.absPath)))
	}

	return fmt.Sprintf("%x", h.Sum(nil))
}

func Of(path string) string {
	info, err := os.Stat(path)
	if err != nil {
		return "NOT_EXISTS"
	}
	if info.IsDir() {
		return directoryDigest(path)
	}
	return fileDigest(path)
}

File: dev-settings/coding-agents/sync-cmd/sync_cmd/checksum.py
import hashlib
from pathlib import Path


def _file_digest(path: Path) -> str:
    try:
        h = hashlib.sha256()
        with open(path, "rb") as f:
            for chunk in iter(lambda: f.read(8192), b""):
                h.update(chunk)
        return h.hexdigest()
    except Exception:
        return "NOT_EXISTS"


def _directory_digest(dir_path: Path) -> str:
    if not dir_path.is_dir():
        return "NOT_EXISTS"

    h = hashlib.sha256()
    entries = sorted(
        (p.relative_to(dir_path), p)
        for p in dir_path.rglob("*")
        if p.is_file()
    )
    for rel, abs_path in entries:
        h.update(str(rel).encode())
        h.update(_file_digest(abs_path).encode())

    return h.hexdigest()


def of(path: Path) -> str:
    path = Path(path)
    if not path.exists():
        return "NOT_EXISTS"
    if path.is_dir():
        return _directory_digest(path)
    return _file_digest(path)
